### PR TITLE
local connection plugin - fix TypeError (#66560)

### LINKED ISSUE #66560: TypeError: Can't mix strings and bytes in path components (on localhost)
##### SUMMARY
Running ansible on localhsot causes a fail: 'Unexpected failure during module execution' (TypeError: Can't mix strings and bytes in path components)

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
/usr/lib/python3.7/site-packages/ansible/modules/setup.py

##### ANSIBLE VERSION

```paste below
ansible 2.9.2
  config file = /home/dborowski/Projects/ansible/localhost/ansible.cfg
  configured module search path = ['/home/dborowski/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 3.7.6 (default, Dec 19 2019, 22:52:49) [GCC 9.2.1 20190827 (Red Hat 9.2.1-1)]
```

##### CONFIGURATION

```paste below
DEFAULT_HOST_LIST(/home/dborowski/Projects/ansible/localhost/ansible.cfg) = ['/home/dborowski/Projects/ansible/localhost/hosts']
DEFAULT_LOCAL_TMP(/home/dborowski/Projects/ansible/localhost/ansible.cfg) = /home/dborowski/Projects/ansible/localhost/tmp/ansible-local-16177azunq0u2
```

##### OS / ENVIRONMENT
 OS: Fedora 31 ThirtyOne
 Kernel: x86_64 Linux 5.4.10-200.fc31.x86_64
 CPU: Intel Core i7-9750H @ 12x 4.5GHz
 GPU: GeForce GTX 1650/PCIe/SSE2
 RAM: 7915MiB / 23965MiB

##### STEPS TO REPRODUCE

ansible-playbook setup-host.yml

```yaml
---
- hosts: localhost
  connection: local

  tasks:
    - name: ensure system is up to date
      become: yes
      dnf:
        name: '*'
        state: latest
```

##### EXPECTED RESULTS

Playbook execution succeeds.

##### ACTUAL RESULTS

Playbook execution fails

```paste below
                             dborowski@localhost
ansible-playbook 2.9.2
  config file = /home/dborowski/Projects/ansible/localhost/ansible.cfg
  configured module search path = ['/home/dborowski/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.7/site-packages/ansible
  executable location = /usr/bin/ansible-playbook
  python version = 3.7.6 (default, Dec 19 2019, 22:52:49) [GCC 9.2.1 20190827 (Red Hat 9.2.1-1)]
Using /home/dborowski/Projects/ansible/localhost/ansible.cfg as config file
setting up inventory plugins
host_list declined parsing /home/dborowski/Projects/ansible/localhost/hosts as it did not pass its verify_file() method
Skipping due to inventory source not existing or not being readable by the current user
script declined parsing /home/dborowski/Projects/ansible/localhost/hosts as it did not pass its verify_file() method
auto declined parsing /home/dborowski/Projects/ansible/localhost/hosts as it did not pass its verify_file() method
Skipping due to inventory source not existing or not being readable by the current user
yaml declined parsing /home/dborowski/Projects/ansible/localhost/hosts as it did not pass its verify_file() method
Skipping due to inventory source not existing or not being readable by the current user
ini declined parsing /home/dborowski/Projects/ansible/localhost/hosts as it did not pass its verify_file() method
Skipping due to inventory source not existing or not being readable by the current user
toml declined parsing /home/dborowski/Projects/ansible/localhost/hosts as it did not pass its verify_file() method
[WARNING]: Unable to parse /home/dborowski/Projects/ansible/localhost/hosts as an inventory source

[WARNING]: No inventory was parsed, only implicit localhost is available

[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not
match 'all'

Loading callback plugin default of type stdout, v2.0 from /usr/lib/python3.7/site-packages/ansible/plugins/callback/default.py

PLAYBOOK: setup-host.yml ********************************************************************************************
Positional arguments: setup-host.yml
verbosity: 4
connection: smart
timeout: 10
become_method: sudo
tags: ('all',)
inventory: ('/home/dborowski/Projects/ansible/localhost/hosts',)
forks: 5
1 plays in setup-host.yml

PLAY [localhost] ****************************************************************************************************

TASK [Gathering Facts] **********************************************************************************************
task path: /home/dborowski/Projects/ansible/localhost/setup-host.yml:2
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: dborowski
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo ./tmp/ansible-tmp-1579247197.7736385-17859363294564 `" && echo ansible-tmp-1579247197.7736385-17859363294564="` echo ./tmp/ansible-tmp-1579247197.7736385-17859363294564 `" ) && sleep 0'
Using module file /usr/lib/python3.7/site-packages/ansible/modules/system/setup.py
The full traceback is:
Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/ansible/executor/task_executor.py", line 146, in run
    res = self._execute()
  File "/usr/lib/python3.7/site-packages/ansible/executor/task_executor.py", line 645, in _execute
    result = self._handler.run(task_vars=variables)
  File "/usr/lib/python3.7/site-packages/ansible/plugins/action/gather_facts.py", line 70, in run
    res = self._execute_module(module_name=fact_module, module_args=mod_args, task_vars=task_vars, wrap_async=False)
  File "/usr/lib/python3.7/site-packages/ansible/plugins/action/__init__.py", line 830, in _execute_module
    self._transfer_data(remote_module_path, module_data)
  File "/usr/lib/python3.7/site-packages/ansible/plugins/action/__init__.py", line 441, in _transfer_data
    self._transfer_file(afile, remote_path)
  File "/usr/lib/python3.7/site-packages/ansible/plugins/action/__init__.py", line 418, in _transfer_file
    self._connection.put_file(local_path, remote_path)
  File "/usr/lib/python3.7/site-packages/ansible/plugins/connection/local.py", line 144, in put_file
    out_path = self._ensure_abs(out_path)
  File "/usr/lib/python3.7/site-packages/ansible/plugins/connection/local.py", line 135, in _ensure_abs
    path = os.path.normpath(os.path.join(self.cwd, path))
  File "/usr/lib64/python3.7/posixpath.py", line 94, in join
    genericpath._check_arg_types('join', a, *p)
  File "/usr/lib64/python3.7/genericpath.py", line 155, in _check_arg_types
    raise TypeError("Can't mix strings and bytes in path components") from None
TypeError: Can't mix strings and bytes in path components

fatal: [localhost]: FAILED! => {
    "msg": "Unexpected failure during module execution.",
    "stdout": ""
}

PLAY RECAP **********************************************************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0
```

### PR DESCRIPTION
**Pull request by heiparta**
##### SUMMARY
Fixes #66560

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
local connection plugin

##### ADDITIONAL INFORMATION

Bug was possibly introduced in https://github.com/ansible/ansible/commit/342d77b32d9176e64ddc89ea886e7560fc56503a

There are no existing unit tests for local connection plugin so just manual verification:
```bash
(venv) ➜  ansible git:(66560) ✗ python --version                                     
Python 3.5.2

(venv) ➜  ansible git:(66560) ✗ cat test_playbook.yaml 
---
- hosts: localhost
  connection: local

  tasks:
    - debug: msg=foobar

# Before fix
(venv) ➜  ansible git:(66560) ✗ ansible-playbook -i inventory.yaml test_playbook.yaml

PLAY [localhost] ******************************************************************************************************************************************************************************

TASK [Gathering Facts] ************************************************************************************************************************************************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: TypeError: Can't mix strings and bytes in path components
fatal: [localhost]: FAILED! => {"msg": "Unexpected failure during module execution.", "stdout": ""}

PLAY RECAP ************************************************************************************************************************************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0  

# After fix
(venv) ➜  ansible git:(66560) ✗ ansible-playbook -i inventory.yaml test_playbook.yaml

PLAY [localhost] ******************************************************************************************************************************************************************************

TASK [Gathering Facts] ************************************************************************************************************************************************************************
ok: [localhost]

TASK [debug] **********************************************************************************************************************************************************************************
ok: [localhost] => {
    "msg": "foobar"
}

PLAY RECAP ************************************************************************************************************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0 

```

### CODE REVIEW

**Comment by heiparta** [comment]
I revised the code to work with self.cwd set as either binary and text, and added a simple unit test

File: lib/ansible/plugins/connection/local.py [hunk 1]
```diff
@@ -132,6 +132,8 @@ def exec_command(self, cmd, in_data=None, sudoable=True):
 
     def _ensure_abs(self, path):
         if not os.path.isabs(path) and self.cwd is not None:
+            if isinstance(self.cwd, binary_type):
+                path = to_bytes(path, errors='surrogate_or_strict')
```
Comment (sivel) [inline lib/ansible/plugins/connection/local.py]: When dealing with paths, it's usually best to ensure you are always dealing with bytes.  Instead of this conditional, the parts should always be ensured as bytes using `to_bytes`.  You may additionally need to validate that the callers of `_ensure_abs` always expect bytes.

**Review by sivel — CHANGES_REQUESTED** [review]

File: lib/ansible/plugins/connection/local.py (hunk 1 above)
Comment (heiparta) [reply lib/ansible/plugins/connection/local.py]: Thanks for the comments!

Fair point about checking what the caller wants. I changed to code to check the type of the input 'path' and make the return value correspond with that. The actual os.path.join will be done with bytes.

**Comment by tymartin-novu** [comment]
For whatever it's worth, the changes proposed in this pull request fixed the `TypeError: Can't mix strings and bytes in path components` issue that we were running into with several of our organization's playbooks.

File: lib/ansible/plugins/connection/local.py [hunk 2]
```diff
@@ -131,8 +131,14 @@ def exec_command(self, cmd, in_data=None, sudoable=True):
         return (p.returncode, stdout, stderr)
 
     def _ensure_abs(self, path):
+        ''' join the path with optional cwd and ensure that the type stays the same '''
         if not os.path.isabs(path) and self.cwd is not None:
-            path = os.path.normpath(os.path.join(self.cwd, path))
+            is_binary = isinstance(path, binary_type)
```
Comment (jamescassell) [inline lib/ansible/plugins/connection/local.py]: Why is this needed?

File: lib/ansible/plugins/connection/local.py [hunk 3]
```diff
@@ -131,8 +131,14 @@ def exec_command(self, cmd, in_data=None, sudoable=True):
...
+            path = to_bytes(path, errors='surrogate_or_strict')
+            cwd = to_bytes(self.cwd, errors='surrogate_or_strict')
+            path = os.path.normpath(os.path.join(cwd, path))
+            if not is_binary:
```
Comment (jamescassell) [inline lib/ansible/plugins/connection/local.py]: Same question.

File: test/units/plugins/connection/test_connection.py [hunk 4]
```diff
@@ -135,7 +137,16 @@ def test_lxc_connection_module(self):
         self.assertIsInstance(LxcConnection(self.play_context, self.in_stream), LxcConnection)
 
     def test_local_connection_module(self):
-        self.assertIsInstance(LocalConnection(self.play_context, self.in_stream), LocalConnection)
+        local = LocalConnection(self.play_context, self.in_stream)
+        self.assertIsInstance(local, LocalConnection)
+
+        # Ensure that file paths are encoded properly
+        tempdir = tempfile.mkdtemp()
+        local.cwd = tempdir.encode('utf-8')  # action plugin sets the cwd as 'bytes'
```
Comment (jamescassell) [inline test/units/plugins/connection/test_connection.py]: Have you bisected the change?

**Review by jamescassell — COMMENTED** [review]
Seems like this would be good to fix right away, but I'd like it to be the correct fix.

File: lib/ansible/plugins/connection/local.py (hunk 2 above)
Comment (heiparta) [reply lib/ansible/plugins/connection/local.py]: Well, the idea is to check the type of the incoming "path" and return the concatenated value in same encoding. In other words, make sure calling _ensure_abs does not unexpectedly change the type of the input parameter if path concatenation is done.

E.g. without this check, the vvv logging in put_file would not look too good, as str path would be converted to bytes by _ensure_abs call.

IMHO the interface of connection plugins (especially LocalConnection) looks bit messy, for example self.cwd is directly modified from outside so there is no control on its encoding.

Anyway, I'm happy to change the implementation if needed. The current version is just the minimal patch to fix the major issue with local playbooks, implemented as safely as I deemed possible.

File: test/units/plugins/connection/test_connection.py (hunk 4 above)
Comment (heiparta) [reply test/units/plugins/connection/test_connection.py]: I'm afraid I don't understand what you mean. This code is an addition to an existing unit test to verify that LocalConnection.put_file now works.

If you were asking if I have bisected from which exact commit the bug originates, then no. I did not check the root cause as in my opinion the fix should anyway be done in _ensure_abs.

**Review by jamescassell — APPROVED** [review]
Thanks for the explanation. Sounds good to me.

shipit

**Comment by heiparta** [comment]
Rebased to migrate the changes to the new test file test_local.py

File: lib/ansible/plugins/connection/local.py (hunk 2 above)
Comment (sivel) [inline lib/ansible/plugins/connection/local.py]: Callers of this method ultimately expect the data to be bytes.  Just ensure that this method returns bytes.  I would start this from the beginning, and the first thing in this function should be `b_path = to_bytes(path, errors='surrogate_or_strict')`.

In addition to that, to self document this, make sure that variables that are bytes are prefixed by `b_`, such as `b_path` and `b_cwd`.  The overhead of calling `to_bytes` on something that is already bytes should be minimal.

**Review by sivel — CHANGES_REQUESTED** [review]

File: lib/ansible/plugins/connection/local.py (hunk 2 above)
Comment (heiparta) [reply lib/ansible/plugins/connection/local.py]: I had somehow omitted the to_native calls in put_file. Therefore I thought the callers are expecting text and implemented the superfluous type checking.

Now all checks have been removed and return type is always bytes.

Comment (heiparta) [reply lib/ansible/plugins/connection/local.py]: @sivel, Can I please get a re-review for the changes?

**Comment by heiparta** [comment]
Closing this PR as the problem was fixed in devel while fixing https://github.com/ansible/ansible/pull/68720
